chain/vm/fvm.go

```diff
@@ -308,6 +315,177 @@ func NewFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
 	}, nil
 }
 
+func NewDebugFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
+	state, err := state.LoadStateTree(cbor.NewCborStore(opts.Bstore), opts.StateBase)
+	if err != nil {
+		return nil, err
+	}
+
+	circToReport, err := opts.CircSupplyCalc(ctx, opts.Epoch, state)
+	if err != nil {
+		return nil, err
+	}
+
+	baseBstore := opts.Bstore
+	overlayBstore := blockstore.NewMemorySync()
+	vmBstore := blockstore.NewTieredBstore(overlayBstore, baseBstore)
+
+	fvmopts := &ffi.FVMOpts{
+		FVMVersion: 0,
+		Externs: &FvmExtern{
+			Rand:       opts.Rand,
+			Blockstore: vmBstore,
+			lbState:    opts.LookbackState,
+			base:       opts.StateBase,
+			epoch:      opts.Epoch,
+		},
+		Epoch:          opts.Epoch,
+		BaseFee:        opts.BaseFee,
+		BaseCircSupply: circToReport,
+		NetworkVersion: opts.NetworkVersion,
+		StateBase:      opts.StateBase,
+		Tracing:        EnableDetailedTracing,
+	}
+
+	fvmopts.Debug = true
+
+	// TODO move these two util functions somewhere more general
```

Comment:
Yes, definitely. If you can, take a look at code blocks [here](https://github.com/filecoin-project/lotus/blob/30981d0fdddac09b7a3ebb7d96e5a911b01eb997/chain/consensus/filcns/upgrades.go#L1483-L1508) and [here](https://github.com/filecoin-project/lotus/blob/30981d0fdddac09b7a3ebb7d96e5a911b01eb997/build/builtin_actors.go#L192-L229). There is likely some de-duping that can happen, I think the right place for this would be `chain/actors/manifest.go`, probably.